experiment: spin locks

hugosenari opened the pull request:
__Disclaimer:__

This is another experiment, I would not expect it to be approved

__Motivation__: Same as from #21

The initial idea was to use spin locks instead of locks, following [this post](https://rigtorp.se/spinlock/) recommendations, I'm using TTAS instead of TAS, I would like to use PAUSE, but is compiler/platform specific.

__Unexpected Outcome:__

It went pretty bad, worse than current implementation.
So I created another version using more spin locks (like I did for #21), one per thread, so they almost won't fight with each other. And that one went very well.

__Results__

We can't compare it with #21 result because I changed the test to set EPOCH after waitAll and before spawn, is more accurate to the test intent (check how much time we are spending between spawn)

The [benchmark results](https://gist.github.com/hugosenari/45580bea0a476a095e82ea06a5adb841),

Since only spin locks makes results worse, so I focus on second implementation _spin\_doctors_, it speeds up the results between 3~5 times.

But still no perfect:
1.  T1E0 - T0E1 increased because T0E1 decreased 75% percentile from ~19767ns, to ~2724ns
2.  OP increased because spin locks uses 100% (of CPU)
    - For this test 1 run with 3 spawns, means 5 others are just spinning.
    - Maybe is why I get better results with more spawns

__Suggestion__:
Some sleep, no stealing.

I know that threads let us sleep using nanoseconds, we could sleep after N runs without any task, this could also help timeout.

There is no task to steal, because I'm only scheduling the same amount of threads as tasks

Araq commented:
There is ` cpuRelax()` and it's mapped to PAUSE or similar, enjoy.

So I think the takeaway is that my stuff is very hard to improve upon? :P

Please note that it is important that eventually there is no busy-loop and the OS is allowed to pause threads as this can be important for energy consumption and Malebolgia is supposed to run well on embedded systems. However, a runtime flag like `useBusyLoop: bool` could be added to the master object.

hugosenari commented on src/experiment/malebolgia_spin_doctors.nim:
```diff
@@ -0,0 +1,309 @@
+    # Test and Set
+    if chan.board[i].compareExchange(empty,
+        tsFeeding, moAcquire, moRelease):
+      ackchyually = i
+      break
```
```suggestion
      ackchyually = i
      signal(chan.alarms[i])
      break
```

To wake up only when empty

hugosenari commented on src/experiment/malebolgia_spin_doctors.nim:
```diff
@@ -0,0 +1,309 @@
+  var
+    item: PoolTask
+    todo = tsToDo
+    wip  = tsWip
+  while not globalStopToken.load(moRelaxed):
+    # Test
+    if chan.board[i].load(moRelaxed) != todo:
+      cpuRelax()
+      continue
```
```suggestion
    wip  = tsWip
    done = tsDone
    idle = 0
  while not globalStopToken.load(moRelaxed):
    # Test
    if chan.board[i].load(moRelaxed) != todo:
      inc idle
      if idle > 1000 and chan.board[i].compareExchange(done, empty, moRelaxed, moRelease):
        idle = 0
        wait(chan.alarms[i])
      cpuRelax()
      continue
```
wait after N idle runs

hugosenari commented on src/experiment/malebolgia_spin_doctors.nim:
```diff
@@ -0,0 +1,309 @@
+
+template spawnImplNoRes(master: var Master; fn: typed) =
+  if stillHaveTime(master):
+    if busyThreads.load(moRelaxed) < ThreadPoolSize:
+      taskCreated master
+      send PoolTask(m: addr(master), t: toTask(fn), result: nil)
+    else:
+      fn
```
I know isn't so simple, but:

If mainThread is used to perform some task, who is there to say to other threads what to do. 
We win one more thread and may lose -  they can be idle - seven.

```suggestion
    if isMainThread() or busyThreads.load(moRelaxed) < ThreadPoolSize:
      taskCreated master
      send PoolTask(m: addr(master), t: toTask(fn), result: nil)
    else:
      fn
```

Araq replied:
That's a good idea.

hugosenari commented:
> There is ` cpuRelax()` and it's mapped to PAUSE or similar, enjoy.

Updated, thanks, 

> So I think the takeaway is that my stuff is very hard to improve upon? :P

You are lazy naming variables, other that, you try to do your best, so there is less room for improvement (in performance) ;-)

> However, a runtime flag like `useBusyLoop: bool` could be added to the master object.

I added suggestions for a wait/signal, but I forgot to review send iterator
It is currently round-robin, means no matter if thread01 is free I set task to next one, this would wake up all threads instead of use ones that are already up.

Round-robin:
- more time with global atom currentThread (x1)
- less likely to test current thread as busy (x8)

Start from 0 always
- no time with global atom currentThread (x1)
- more likely to test current thread as busy (x8)
- but won't wake all threads

hugosenari commented on bench/parapply/tparapply_male.nim:
```diff
@@ -0,0 +1,113 @@
+  bigbang = getMonoTime()
+  parApply ops, 2, now
+  
+  echo [
+    $inNanoseconds(epoch  - bigbang),  # mesurement precision
+    $inNanoseconds(ops[0] - epoch),    # how fast we perform 1º task
+    $inNanoseconds(ops[1] - epoch),    # how fast we perform 2º task serial after 1º
+    $inNanoseconds(ops[2] - epoch),    # how fast we perform 3º task parallel
+    $inNanoseconds(ops[3] - epoch),    # how fast we perform 4º task serial after 3º
+    $inNanoseconds(ops[4] - epoch),    # how fast we perform 5º task parallel
+    $inNanoseconds(ops[1] - ops[0]),   # serial   latency
+    $inNanoseconds(ops[3] - ops[1]),   # parallel latency
+  ].join(sep)
```
This answers your question about benchmark results
```nim
echo [
  "OP",
  "T0E0",
  "T0E1",
  "T1E0",
  "T1E1",
  "T2E0",
  "T0E1-T0E0",
  "T1E0-T0E1",
].join(sep)
```

Now I realized that OP is misaligned, it was delta of two consecutive getMonoTime() like T0E1-T0E0 (ops[1]-ops[0]) that has the same effect
Now it is time before parApply until first line of body in waitAll

hugosenari commented:
Closing as experiment completed, 

Next one, should be work stealing, but you already played with it, and mention as slower.
So I am not motivated.

> I'm not afraid of atomic instructions.

I couldn't say the same, I'm still on page 183, and parallelism is only in page 253.

Araq commented:
> Next one, should be work stealing, but you already played with it, and mention as slower. So I am not motivated.

But I didn't use a lockfree queue for it and only run this on a 8 core M1. Work stealing scales much better for bigger hardware. It shouldn't be hard to beat the existing implementation.

hugosenari commented:
> But I didn't use a lockfree queue for it and only tested it on a 8 core M1. Work stealing scales much better for bigger hardware. It shouldn't be hard to beat the existing implementation.

Malebolgia as MASTER branch or this PR in malebolgia_spin.nim

![image](https://github.com/Araq/malebolgia/assets/863299/d1fce768-0b83-43d2-9a8e-c69d76d210f8)

Malebolgia as this PR in malebolgia_spin_doctors.nim

![image](https://github.com/Araq/malebolgia/assets/863299/ff4eca40-4ece-488a-b0bb-dc24029775b9)

My view of how Work Stealing looks like,

![image](https://github.com/Araq/malebolgia/assets/863299/66c464b8-e664-4026-a31a-825e9a6b7197)

...
NOPE! :rofl: 

Ok, to be fair, this representation is worse than real the worst scenario at some point in time:  
Because if a thread is ~shifting~ poping from other it isn't shifting from itself, or all other at the same time.
But for other graphs, I represented a superposition of all possible cases. :thinking:


Notes: 
Simplified to single producer/N consumers, in reality can have N producers/consumers
I know that _.L_ isn't the queue, but is what we have to wait to operate the queue.

Araq commented:
How did you make these graphs? They are awesome!

However, as I said, work stealing doesn't even have to use locks at all and can be done with a lockfree queue. And more importantly, all the highly concurrent/parallel runtimes ended up doing work stealing so it's reasonable to assume that it simply is a very good mechanism.

hugosenari commented:
> How did you make these graphs? They are awesome!

Coggle.it

https://coggle.it/diagram/ZQDcj9E4196r1jqQ/t/t0
https://coggle.it/diagram/ZQFMXn170vn6vXB5/t/t0-2023-9-12-22-39
https://coggle.it/diagram/ZPKLGeut86Ir78iB/t/t0
